### ISSUE TITLE
Consul connect: Consul agent fails to renew leaf certificate

### ISSUE THREAD

**Issue by thanapolr** [hashicorp/consul#5239]
Reproducing Steps:
- Create 3 consul 1.4.0 servers, 3 vault 1.0.2 servers, 2 consul 1.4.0 agents using vagrant
- enable consul connect on server using consul CA or vault CA

-- configuration for using consul CA
```
connect {
    enabled = true
}
```
- setup consul agent-1 with the following configuration to register service with sidecar proxy and upstreams
```
services = [
    {
        id = "dummy1"
        name = "dummy1"
        port = 8080
        connect {
            sidecar_service {
                proxy {
                    upstreams = [
                        {
                            destination_name = "dummy2"
                            local_bind_port = 10002
                        }
                    ]
                }
            }
        }
    }
]
```
- setup consul agent-2 with the following configuration to register service with sidecar proxy and upstreams
```
services = [
    {
        id = "dummy2"
        name = "dummy2"
        port = 8080
        connect {
            sidecar_service {
                proxy {
                    upstreams = [
                        {
                            destination_name = "dummy1"
                            local_bind_port = 10001
                        }
                    ]
                }
            }
        }
    }
]
```
- Both agents have acl token with service write policy since we use master token(the issue exists even with other tokens with service write policy) and is configured as following : 
```
{
  "primary_datacenter": "local",
  "acl": {
    "enabled": true,
    "tokens": {
      "default": "{{ consul-master-token }}"
    }
  }
}
```
- When the consul agent-1 and agent-2 are started, they receive valid leaf certificates and can communicate through proxies using mTLS.
- Now if we change the date on the agent-1 to make the certificate expire, or if you just let it expire eventually, then it couldn't get/renew a new valid leaf certificate and the services couldn't communicate anymore. The following is the error in the logs:

```
[ERR] http: Request GET /v1/agent/connect/ca/leaf/dummy1?index=11992410, error: No known Consul servers from=127.0.0.1:43086
```

**Comment by elocnatsirt** [comment]
I am seeing this same error using the Vault CA and Connect Injector for K8s. After ~3.65 days, services are no longer able to communicate with each other and require a full restart/new injection to function as a Connect service.

Is there a fix for this?

**Comment by t0nyhays** [comment]
> I am seeing this same error using the Vault CA and Connect Injector for K8s. After ~3.65 days, services are no longer able to communicate with each other and require a full restart/new injection to function as a Connect service.
> 
> Is there a fix for this?

I had been having this issue with Connect and Consul CA since initial release.  1.4.2 has fixed it for me.  Have been running around the clock since install ~ 1 month.  Prior I had never been longer than a week.

**Comment by elocnatsirt** [comment]
> I had been having this issue with Connect and Consul CA since initial release. 1.4.2 has fixed it for me. Have been running around the clock since install ~ 1 month. Prior I had never been longer than a week.

After upgrading the servers to 1.4.4, so far everything seems to be running smoothly and I am not getting the errors after a few days using the Vault CA. Will report back if I notice any issues, but upgrading seemed to have fixed this for me.

**Comment by i0rek** [comment]
Thank you everybody for reporting and updating this issue. Is anybody still experiencing this issue with Consul >= 1.4.2?

**Comment by kyhavlov** [comment]
After investigating this it seems like this issue was fixed by #4480, which was one of a few different fixes around leaf certificates in 1.4.1. There were a couple other subtle issues that were possible to hit as well that could exacerbate the situation seen here as well. I'm going to close this issue as this problem is fixed by newer Consul versions, but if anyone still sees this behavior on a newer version (>= 1.4.2) it'd be worth opening a new issue since that would probably have a different underlying cause.
